Add kube-proxy non-privileged blog post (#452)
kubernetes/contributor-site: content/en/blog/2023/0000-00-00-kube-proxy-non-privileged.md

---
layout: blog

> title: "Kubernetes 1.29: Support for Running kube-proxy in a Non-privileged Container"

Reviewer: We ran out of capacity to review and publish this as part of the post-release blog series.

How about:
```suggestion
title: "Kubernetes supports running kube-proxy in an unprivileged container"
```
?

> date: 0000-00-00

Reviewer: ```suggestion
date: 2024-01-05
```
maybe?

slug: kube-proxy-non-privileged
---

> **Author**: Lars Ekman

Reviewer: For this site, the author information goes in the front matter. See https://github.com/kubernetes/contributor-site/blob/118088ef5115b4665c7b3ce6ee32579dbd494cf7/content/en/blog/2023/contextual-logging-in-kubernetes-1-29.md?plain=1 for an example.

This post describes how the `--init-only` flag to `kube-proxy` can be
used to run the main kube-proxy container in a stricter

> `securityContext` by performing the configuration that requires

Reviewer: ```suggestion
`securityContext`, by performing the configuration that requires
```

> privileged mode in a separate [init container](
> /docs/concepts/workloads/pods/init-containers/). Since

Reviewer: ```suggestion
privileged mode in a separate [init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/).
Since
```

Windows doesn't have the equivalent of `capabilities`, this only works
on Linux.

The `kube-proxy` Pod still only meets the *privileged* [Pod Security
Standard](https://kubernetes.io/docs/concepts/security/pod-security-standards/),
but there is still an improvement because the running container doesn't
need to run privileged.

Please note that `kube-proxy` can be installed in different ways. The
examples below assume that kube-proxy is run from a pod, but similar
changes could be made in clusters where it is run as a system service.

## Background

It is undesirable to run a server container like `kube-proxy` in
privileged mode. Security aware users wants to use capabilities instead.

If `kube-proxy` is installed as a POD, the initialization requires
"privileged" mode, mostly for setting sysctl's. However, `kube-proxy`
only tries to set the sysctl's if they don't already have the right

> values. In theory, then, if a privileged [init container](/docs/concepts/workloads/pods/init-containers/)

Reviewer: ```suggestion
values. In theory, then, if a privileged [init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
```

or omit the link, as we link to it earlier in the article.

made, but with `--init-only` you can have `kube-proxy` itself do the setup
*exactly* as on a normal start, and then exit.

## Initializing kube-proxy in an init container

*The example manifests below are not complete, but narrowed down to what is
essential to illustrate the function.*

Usually, cluster operators run `kube-proxy` in a privileged security context.

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    k8s-app: kube-proxy
spec:
  template:
    spec:
      containers:
      - name: kube-proxy
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        securityContext:
          privileged: true
   # (lots of stuff omitted here...)
```

But now it is possible to use:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    k8s-app: kube-proxy
spec:
  template:
    spec:
      initContainers:
      - name: kube-proxy-init
        command:
        - /usr/local/bin/kube-proxy
        - --config=/var/lib/kube-proxy/config.conf
        - --hostname-override=$(NODE_NAME)
        - --init-only
        securityContext:
          privileged: true

>    # (lots of stuff omitted here...)

Reviewer: nice!
